File: src/collectors/editors-ext.ts
import type { Collector, CollectorResult } from "./types";
import { makeSection } from "./types";
import { type CommandEnv, defaultEnv } from "./env";

/** Parse `code --list-extensions` / `cursor --list-extensions` (one id per line). */
export function parseExtensions(text: string): string[] {
  return text
    .trim()
    .split("\n")
    .map((l) => l.trim())
    .filter(Boolean)
    .sort();
}

const items = (exts: string[]) => exts.map((e) => ({ raw: e, columns: [e] }));

export function makeEditorsExtCollector(env: CommandEnv = defaultEnv): Collector {
  return async () => {
    const result: CollectorResult = {};

    try {
      const code = parseExtensions(await env.run(["code", "--list-extensions"]));
      if (code.length) result["editor.vscode.extensions"] = makeSection("editor.vscode.extensions", { items: items(code) });
    } catch {}

    try {
      const cursor = parseExtensions(await env.run(["cursor", "--list-extensions"]));
      if (cursor.length) result["editor.cursor.extensions"] = makeSection("editor.cursor.extensions", { items: items(cursor) });
    } catch {}

    return result;
  };
}

export const collectEditorsExt = makeEditorsExtCollector();

File: internal/collect/editorsext.go
package collect

import (
	"sort"
	"strings"

	"github.com/doguyilmaz/dothaven/internal/snapshot"
)

// ParseExtensions parses `code --list-extensions` / `cursor --list-extensions`
// output (one extension id per line): trims, drops blanks, and sorts.
func ParseExtensions(text string) []string {
	lines := strings.Split(strings.TrimSpace(text), "\n")
	out := make([]string, 0, len(lines))
	for _, l := range lines {
		l = strings.TrimSpace(l)
		if l != "" {
			out = append(out, l)
		}
	}
	sort.Strings(out)
	return out
}

// EditorsExtCollector lists installed VS Code and Cursor extensions.
func EditorsExtCollector(c Ctx) snapshot.Snapshot {
	out := snapshot.Snapshot{}

	editors := []struct {
		cmd     string
		section string
	}{
		{"code", "editor.vscode.extensions"},
		{"cursor", "editor.cursor.extensions"},
	}
	for _, e := range editors {
		stdout, _ := c.Env.Run(c.Context, e.cmd, "--list-extensions")
		exts := ParseExtensions(stdout)
		if len(exts) > 0 {
			out[e.section] = snapshot.Section{Items: toItems(exts)}
		}
	}

	return out
}
